app/src/main/java/org/mozilla/fenix/components/metrics/AppLaunchTimeMeasurement.kt

```diff
@@ -0,0 +1,92 @@
+import kotlinx.coroutines.withContext
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.HOT
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.WARM
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.COLD
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.UNKNOWN
+import org.mozilla.fenix.perf.Stat
+
+/**
+ * Handles the logic of figuring out launch time for cold, warm and hot startup.
+ */
+class AppLaunchTimeMeasurement(private val stats: Stat = Stat()) {
+
+    private var isOnPreDrawCalled = false
+    private var startupMode = UNKNOWN
+
+    private var applicationOnCreateTimeStampNanoSeconds: Long? = null
+    private var homeActivityInitTimeStampNanoSeconds: Long? = null
+    private var homeActivityOnRestartTimeStampNanoSeconds: Long? = null
+    // we are considering screen to be visible after the first pre draw call.
+    private var homeActivityOnPreDrawTimeStampNanoSeconds: Long = INVALID_TIMESTAMP
+
+    fun onApplicationOnCreate() {
+        startupMode = COLD
+    }
+
+    fun onHomeActivityOnCreate(activityInitNanos: Long) {
+        if (startupMode == UNKNOWN) {
+            startupMode = WARM
+        }
+        this.homeActivityInitTimeStampNanoSeconds = activityInitNanos
+    }
+
+    fun onHomeActivityOnRestart(activityOnRestartNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        startupMode = HOT
+        homeActivityOnRestartTimeStampNanoSeconds = activityOnRestartNanos
+    }
+
+    fun onActivityOnPreDraw(activityOnPreDrawNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        isOnPreDrawCalled = true
+        homeActivityOnPreDrawTimeStampNanoSeconds = activityOnPreDrawNanos
+    }
+
+    /**
+     * if we have both start and finish time for launch, return the difference otherwise return null.
+     */
+    suspend fun getApplicationLaunchTime(): Long? = withContext(Dispatchers.IO) {
+        when {
+            !isOnPreDrawCalled -> {
+                null
+            }
+            else -> {
+                when (startupMode) {
+                    COLD -> {
+                        applicationOnCreateTimeStampNanoSeconds =
+                            stats.getProcessStartTimeStampNano(Process.myPid())
+                        homeActivityOnPreDrawTimeStampNanoSeconds.minus(
+                            applicationOnCreateTimeStampNanoSeconds!!
+                        )
+                    }
+                    WARM -> {
+                        homeActivityOnPreDrawTimeStampNanoSeconds.minus(
+                            homeActivityInitTimeStampNanoSeconds!!
+                        )
+                    }
+                    HOT -> {
+                        homeActivityOnPreDrawTimeStampNanoSeconds.minus(
+                            homeActivityOnRestartTimeStampNanoSeconds!!
+                        )
+                    }
+                    else -> {
```

Comment:
-> UNKNOWN

If you have an exhaustive type such as an enum, you should specify all of them instead of using `else`: using non-else values, the compiler will warn you if you don't handle all the cases, which is helpful if you later add a new case and forget to handle it everywhere.

Reply:
done!